src/Microsoft.ML.SearchSpace/SearchSpace.cs

```diff
@@ -0,0 +1,386 @@
+
+        public override int?[] Step
+        {
+            get
+            {
+                return this.OrderBy(kv => kv.Key)
+                           .SelectMany(kv => kv.Value.Step)
+                           .ToArray();
+            }
+        }
+
+        public OptionBase this[string key] { get => ((IDictionary<string, OptionBase>)_options)[key]; set => ((IDictionary<string, OptionBase>)_options)[key] = value; }
+
+        public override Parameter SampleFromFeatureSpace(double[] feature)
+        {
+            Contracts.Check(feature.Length == this.FeatureSpaceDim, "input feature doesn't match");
+            var param = Parameter.CreateNestedParameter();
+            var cur = 0;
+
+            foreach (var key in this._options.Keys.OrderBy(k => k))
+            {
+                var option = this._options[key];
+                var input = feature.Skip(cur).Take(option.FeatureSpaceDim).ToArray();
+                var value = option.SampleFromFeatureSpace(input);
+                param[key] = value;
+                cur += option.FeatureSpaceDim;
+            }
+
+            if (this._defaultOption != null)
+            {
+                return this.Update(this._defaultOption, param);
+            }
+
+            return param;
+        }
+
+        public override double[] MappingToFeatureSpace(Parameter parameter)
+        {
+            var res = new List<double>();
+            foreach (var key in this._options.Keys.OrderBy(k => k))
+            {
+                var option = this._options[key];
+                var input = parameter[key];
+                var value = option.MappingToFeatureSpace(input);
+                res.AddRange(value);
+            }
+
+            return res.ToArray();
+        }
+
+        private Dictionary<string, OptionBase> GetOptionsFromType(Type typeInfo)
+        {
+            var fieldOptions = this.GetOptionsFromField(typeInfo);
+            var propertyOptions = this.GetOptionsFromProperty(typeInfo);
+            return fieldOptions.Concat(propertyOptions).ToDictionary(kv => kv.Key, kv => kv.Value);
+        }
+
+        private Dictionary<string, SearchSpace> GetNestedSearchSpaceFromType(Type typeInfo)
+        {
+            var fieldSS = this.GetSearchSpacesFromField(typeInfo);
+            var propertySS = this.GetSearchSpacesFromProperty(typeInfo);
+            return fieldSS.Concat(propertySS).ToDictionary(kv => kv.Key, kv => kv.Value);
+        }
+
+        private Dictionary<string, SearchSpace> GetSearchSpacesFromField(Type typeInfo)
+        {
+            var fieldInfos = typeInfo.GetFields(System.Reflection.BindingFlags.Public | System.Reflection.BindingFlags.Instance);
+            var res = new Dictionary<string, SearchSpace>();
+
+            foreach (var field in fieldInfos)
+            {
+                var optionAttribute = field.GetCustomAttributes(typeof(OptionAttribute), false);
+                if (optionAttribute.Count() == 0)
```

Comment:
Don't use `.Count()`, almost always it is the wrong thing to use. Use `.Length` here instead.

Reply:
### Resolved